src/libraries/System.Private.CoreLib/src/System/Reflection/ConstructorInvoker.cs

```diff
@@ -2,25 +2,94 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.Reflection
 {
     internal sealed partial class ConstructorInvoker
     {
-        private readonly RuntimeConstructorInfo _constructorInfo;
-        public InvocationFlags _invocationFlags;
+        private readonly RuntimeConstructorInfo _method;
+
+#if !MONO // Temporary until Mono is updated.
+        private bool _invoked;
+        private bool _strategyDetermined;
+        private InvokerEmitUtil.InvokeFunc? _emitInvoke;
+#endif
 
         public ConstructorInvoker(RuntimeConstructorInfo constructorInfo)
         {
-            _constructorInfo = constructorInfo;
+            _method = constructorInfo;
+
+#if USE_NATIVE_INVOKE
+            // Always use the native invoke; useful for testing.
+            _strategyDetermined = true;
+#elif USE_EMIT_INVOKE
+            // Always use emit invoke (if IsDynamicCodeCompiled == true); useful for testing.
+            _invoked = true;
+#endif
         }
 
+#if MONO // Temporary until Mono is updated.
+        [DebuggerStepThrough]
+        [DebuggerHidden]
+        public unsafe object? InvokeUnsafe(object? obj, Span<object?> args, BindingFlags invokeAttr)
+        {
+            return InvokeNonEmitUnsafe(obj, args, invokeAttr);
+        }
+#else
         [DebuggerStepThrough]
         [DebuggerHidden]
-        public unsafe object? InvokeUnsafe(object? obj, IntPtr* args, Span<object?> argsForTemporaryMonoSupport, BindingFlags invokeAttr)
+        public unsafe object? InvokeUnsafe(object? obj, IntPtr* args, BindingFlags invokeAttr)
         {
-            // Todo: add strategy for calling IL Emit-based version
-            return _constructorInfo.InvokeNonEmitUnsafe(obj, args, argsForTemporaryMonoSupport, invokeAttr);
+            if (!_strategyDetermined)
```

Comment:
I think we can save some cycles on the golden when we have the delegate already and the exception wrapping is not required. The delegate call can be inlined into the caller in this case. Something like:

```
[AggressiveInlining]
InlinedInvoke()
{
     if (_invokeFunc != null && (invokeAttr & BindingFlags.DoNotWrapExceptions) == 0) && obj == null)
    {
        return _emitFunc(...);
    }
    return Invoke();
}

Invoke()
{
    ... the existing method ...
}
```

Reply:
I assume you meant `!=` instead of `==` in `&& (invokeAttr & BindingFlags.DoNotWrapExceptions) != 0)`? 

Changing that makes sense for the minority that specifies `BindingFlags.DoNotWrapExceptions`. And I do think we should be encouraging `DoNotWrapExceptions` since it usually aids diagnosis plus the semanitcs will be consistent with the future APIs planned for byref-like support.

Reply:
> I assume you meant != instead of == in && (invokeAttr & BindingFlags.DoNotWrapExceptions) != 0)?

Yep.